liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,445 @@
-                        ref n if n == &end_tag && nesting_depth == 0 => break,
-                        _ => {}
-                    }
-                };
-                children.push(t.clone())
-            }
-            options.blocks[x.as_str()].parse(x, &tokens[1..], &children, options)
-        }
+    let mut filter = filter.into_inner();
+    let name = filter.next().expect("A filter always has a name.").as_str();
+    let args = filter.map(parse_value).collect();
 
-        ref x => Err(Error::with_msg("Tag is not supported").context("tag", format!("{}", x))),
-    }
+    FilterCall::new(name, args)
 }
 
-/// Confirm that the next token in a token stream is what you want it
-/// to be. The token iterator is moved to the next token in the stream.
-pub fn expect<'a, T>(tokens: &mut T, expected: &Token) -> Result<&'a Token>
-where
-    T: Iterator<Item = &'a Token>,
-{
-    match tokens.next() {
-        Some(x) if x == expected => Ok(x),
-        x => Err(unexpected_token_error(&format!("`{}`", expected), x)),
+fn parse_filter_chain(chain: Pair) -> FilterChain {
+    if chain.as_rule() != Rule::FilterChain {
+        panic!("Expected an expression with filters.");
     }
+
+    let mut chain = chain.into_inner();
+    let entry = parse_value(
+        chain
+            .next()
+            .expect("A filterchain always has starts by a value."),
+    );
+    let filters = chain.map(parse_filter).collect();
+
+    FilterChain::new(entry, filters)
 }
 
-/// Extracts a token from the token stream that can be used to express a
-/// value. For our purposes, this is either a string literal, number literal
-/// or an identifier that might refer to a variable.
-pub fn consume_value_token(tokens: &mut Iter<Token>) -> Result<Token> {
-    match tokens.next() {
-        Some(t) => value_token(t.clone()),
-        None => Err(unexpected_token_error(
-            "string | number | boolean | identifier",
-            NOTHING,
-        )),
+fn parse_tag<'a>(
+    tag: Pair<'a>,
+    next_elements: &mut Iterator<Item = Pair<'a>>,
+    options: &LiquidOptions,
+) -> Result<Box<Renderable>> {
+    if tag.as_rule() != Rule::Tag {
+        panic!("Expected a tag.");
     }
-}
 
-/// Recognises a value token, returning an error if a non-value token
-/// is presented.
-pub fn value_token(t: Token) -> Result<Token> {
-    match t {
-        v @ Token::StringLiteral(_)
-        | v @ Token::IntegerLiteral(_)
-        | v @ Token::FloatLiteral(_)
-        | v @ Token::BooleanLiteral(_)
-        | v @ Token::Identifier(_) => Ok(v),
-        x => Err(unexpected_token_error(
-            "string | number | boolean | identifier",
-            Some(&x),
-        )),
+    let mut tag = tag.into_inner();
+    let name = tag
+        .next()
+        .expect("A tag starts with an identifier.")
+        .as_str();
+    let mut tokens = tag.map(TagToken::from);
+
+    if options.tags.contains_key(name) {
+        options.tags[name].parse(name, &mut tokens, options)
```

Comment:
Why is tokens passed as a `&mut` rather than transferring ownership?

Reply:
If I remember well, the function signature accepts an `Iterator<Item=TagToken>` as the parameter, which is a trait. This meant that in order to pass a concrete object, I would need instead to pass a pointer (in this case, mutable reference) to it. This probably may be fixed by changing the signature to generics.

Reply:
Ah, good point.

Tags/blocks can't accept generics because they need to be Trait Objects.

One alternative is to pass in a slice instead.  Not sure which would be better.

Reply:
Passing a slice would require to collect the iterator and most tags would probably just turn the slice into an iterator again.

I guess having iterators would be the best option, at least for now.

Reply:
Besides, slices are still tecnically just references so ownership is not transferred.

Reply:
So maybe I should step back and explain what felt weird to me with the API.  To be clear, this is to express my reasoning  and not to say which way it has to be done.

> Besides, slices are still tecnically just references so ownership is not transferred.

A `&mut` is ambiguous about where in the iterator you are starting and where you are ending ie what is the expectation for how each side of the API works.

A `&` slice would better convey "everything in this is for you" (unless we did other stuff like having the return value include the slice).

Reply:
What about using a `Box`? Would that better express the meaning of the API?

Reply:
Aparently that raises problems because `Box` requires a `'static` lifetime.

Reply:
I keep getting this error message
```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'i` due to conflicting requirements
   --> liquid-compiler\src\parser.rs:187:23
    |
187 |     let mut tag = tag.into_inner();
    |                       ^^^^^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the function body at 178:14...
   --> liquid-compiler\src\parser.rs:178:14
    |
178 | fn parse_tag<'a>(
    |              ^^
    = note: ...so that the types are compatible:
            expected pest::iterators::Pair<'_, _>
               found pest::iterators::Pair<'a, _>
    = note: but, the lifetime must be valid for the static lifetime...
    = note: ...so that the expression is assignable:
            expected std::boxed::Box<(dyn std::iter::Iterator<Item=parser::TagToken<'_>> + 'static)>
               found std::boxed::Box<dyn std::iter::Iterator<Item=parser::TagToken<'_>>>
```

Reply:
I think I reproduced your problem on the playground and after doing some searching, I found the part of lifetimes I frequently forget

`Box<Iterator<Item=&'a i32>>` is implicitly `Box<Iterator<Item=&'a i32> + 'static>`, requiring the type you are boxing to be `'static`.  Since your iterator is not, you need to mention the lifetime that the iterator is valid for: `Box<Iterator<Item=&'a i32> + 'a>`

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=37b360c69c4e60658d3012309a1b341e

Reply:
The place I found the answer: https://www.reddit.com/r/rust/comments/325u5a/right_way_to_return_an_iterator_over_a_slice/?st=johq7l62&sh=ea03b4f9

Reply:
Thanks! I had never seen that syntax `+ 'lifetime` before.
I tried to implement that, which led to using a lot of `<'a>` lifetimes in `tag.rs`. Still, there are errors I'm not being able to overcome.
```
error[E0477]: the type `tag::FnTagParser<'a>` does not fulfill the required lifetime
  --> liquid-compiler\src\tag.rs:54:10
   |
54 | impl<'a> ParseTag<'a> for FnTagParser<'a> {
   |          ^^^^^^^^^^^^
   |
   = note: type must satisfy the static lifetime

error[E0477]: the type `tag::BoxedTagParser<'a>` does not fulfill the required lifetime
  --> liquid-compiler\src\tag.rs:76:10
   |
76 | impl<'a> ParseTag<'a> for BoxedTagParser<'a> {
   |          ^^^^^^^^^^^^
   |
   = note: type must satisfy the static lifetime

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0477`.
```

I am very confused, as I don't understand why those would be required to satisfy the static lifetime.

Reply:
Threads, you say? That makes a lot of sense actually, as `ParseTag` implements `Send + Sync`. Maybe the best thing to do right now is to avoid all these changes and leave the iterator as a `&mut` or slice and think about boxing it in a future PR.

Reply:
I agree.  Better to get an imperfect solution in now and then iterate on it.